server/remote_cache/byte_stream_server/byte_stream_server.go

```diff
@@ -402,6 +394,59 @@ func (s *ByteStreamServer) QueryWriteStatus(ctx context.Context, req *bspb.Query
 	}, nil
 }
 
+func handleAlreadyExists(stream bspb.ByteStream_WriteServer, firstRequest *bspb.WriteRequest) error {
+	r, err := digest.ParseUploadResourceName(firstRequest.ResourceName)
+	if err != nil {
+		return err
+	}
+	// The following code implements a workaround for the way Bazel 5.0.0 handles
+	// writing compressed blobs.
+	//
+	// When the server closes the stream and sends the WriteResponse, Bazel will
+	// check whether the committed_size matches the number of bytes it has
+	// uploaded, but only if it has uploaded all bytes. But because Bazel doesn't
+	// wait for an ack before it uploads each chunk, we (generally) have no way of
+	// knowing whether Bazel has uploaded all bytes at this point. So we are
+	// forced to wait until bazel uploads all chunks, so that we know for sure
+	// that the committed_size we send back will match Bazel's expectation.
+	//
+	// There are two cases where we can short-circuit, though:
+	//
+	// - If this is an uncompressed stream, we assume that the committed_size will
+	//   match the digest size. (If the two sizes actually do mismatch, it's more
+	//   likely a Bazel problem, since we implement cache checksums).
+	// - If the first request was marked with FinishWrite, we don't expect
+	//   Bazel to send any more messages.
+	//
+	// See https://github.com/bazelbuild/bazel/issues/14654 for more context
+	committedSize := r.GetDigest().GetSizeBytes()
+	if r.GetCompressor() != repb.Compressor_IDENTITY {
+		remainingSize := int64(0)
+		if !firstRequest.FinishWrite {
+			remainingSize, err = remainingUploadSize(stream)
+			if err != nil {
+				return err
+			}
+		}
+		committedSize = int64(len(firstRequest.Data)) + remainingSize
+	}
+	return stream.SendAndClose(&bspb.WriteResponse{CommittedSize: committedSize})
+}
+
+func remainingUploadSize(stream bspb.ByteStream_WriteServer) (int64, error) {
```

Comment:
nit: comment what this function does?

Reply:
Done, and renamed to make it more clear that it's doing IO